Sources/Fullscreen/ExploreDetail/ExploreDetailView.swift

```diff
@@ -0,0 +1,325 @@
+                ofKind: UICollectionView.elementKindSectionHeader
+            )
+            view.configure(withText: title, font: section.headerFont)
+            return view
+        }
+
+        return dataSource
+    }()
+
+    // MARK: - Init
+
+    public init(dataSource: ExploreDetailViewDataSource, delegate: ExploreDetailViewDelegate) {
+        self.dataSource = dataSource
+        self.delegate = delegate
+
+        super.init(frame: .zero)
+        setup()
+    }
+
+    public required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: - Setup
+
+    public func configure(with viewModel: ExploreDetailViewModel) {
+        heroView.configure(withTitle: viewModel.title, subtitle: viewModel.subtitle, imageUrl: viewModel.imageUrl)
+        heroView.isHidden = !viewModel.showHeroView
+        collectionView.contentInset.top = viewModel.showHeroView ? 220 : 0
+    }
+
+    public func reloadSections() {
+        var snapshot = NSDiffableDataSourceSnapshot<Section, Item>()
+        snapshot.appendSections(sections)
+
+        for section in sections {
+            switch section.items {
+            case .collections(let collections):
+                snapshot.appendItems(collections.map(Item.collection), toSection: section)
+            case .selectedCategories(let collections):
+                snapshot.appendItems(collections.map(Item.selectedCategory), toSection: section)
+            case .ads(let ads):
+                snapshot.appendItems(ads.map(Item.ad), toSection: section)
+            }
+        }
+
+        collectionDataSource.apply(snapshot, animatingDifferences: true)
+    }
+
+    public func updateFavoriteStatusForVisibleItems() {
+        for indexPath in collectionView.indexPathsForVisibleItems {
+            let section = sections[indexPath.section]
+
+            switch section.items {
+            case .collections, .selectedCategories:
+                break
+            case .ads(let items):
+                if let cell = collectionView.cellForItem(at: indexPath) as? ExploreAdCell {
+                    cell.isFavorite = items[indexPath.item].isFavorite
+                }
+            }
+        }
+    }
+
+    private func setup() {
+        backgroundColor = .bgPrimary
+        addSubview(collectionView)
+        addSubview(heroView)
+
+        collectionView.fillInSuperview()
+
+        NSLayoutConstraint.activate([
+            heroViewTopConstraint,
+            heroView.leadingAnchor.constraint(equalTo: leadingAnchor),
+            heroView.trailingAnchor.constraint(equalTo: trailingAnchor),
+            heroViewHeightConstraint
+        ])
+
+        heroView.remoteImageViewDataSource = self
+    }
+}
+
+// MARK: - UICollectionViewDelegate
+
+extension ExploreDetailView: UICollectionViewDelegate {
+    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        let section = sections[indexPath.section]
```

Comment:
Since you are using snapshots, it is recommended to get the item at a specific index path by using 
`dataSource.itemIdentifier(for: indexPath) as? <Item>`, instead of using the array and index. 
The reason we don't want to use the array as a source of truth, is that it can be out of sync with what's actually in the view. The snapshot will _always_ be correct, thus you can't get any index out of bounds crashes, even if doing a lot of operations at the same time.

I'm not sure what you use the index path for in the delegate though, maybe not possible to discard the index in this case? 😄

Reply:
Good point! But I think using snapshots ads more confusion here since the actual source of truth is the array from `func sections(inExploreDetailView view: ExploreDetailView) -> [ExploreDetailSection]` from `ExploreDetailViewDataSource`.  It's implemented this way to be able to reload favorite button without reloading the cell. I can refactor this view without diffable data sources, it's not going to prevent index out of bounds crashes, but at least it should make the code easier to reason about. What do you think @ninarg ?

Reply:
I understand. If this is only needed for the favorite functionality, I think it can be fixed by letting the main app check against the ad's adId in the view model's `isFavorite` getter. We have done something similar for search quest. That way it should be possible to update favorite states this way: 
```
for cell in collectionView.visibleCells {
       if let cell = cell as? ExploreAdCell {
                cell.updateFavoriteState() 
```
where the cell just updates the fave button with `viewmodel.isFavorite`. I might be mistaken, but I think it would be possible to use the snapshot directly with this approach 🤔 

It's just a proposed solution for when you might have time to look into it.  What we have now works fine, also with the diffable data source, so I will accept it as is. 😄